### ISSUE TITLE
It may be possible for Http to run JS on the wrong thread

### ISSUE THREAD

**Issue by cbaggers** [fusetools/fuselibs-public#641]
In `Source/Fuse.Scripting.JavaScript/FuseJS/Http.uno` there are many calls to `Scripting.Function` objects without a clear guarentee that they only happen on the language thread

I have not been able to nail down the behaviour here but I have not yet found a place where Invoke is used so this is suspect and should be clarified. Use of `Scripting.Context` would make the issue very explicit.

**Comment by cbaggers** [comment]
warning: hypothosis, havent checked the code properly yet, I just saw this whilst fixing something else and wanted to note it down

```
return new FuseJSHttpRequest(_context, _client.CreateRequest(method, url, _context.Dispatcher)).Obj;
```
I think this is the reason that the JS may be run on the correct thread, namely we pass in an `IDispatcher` that dictates what threads the resulting events are fired on.

**Comment by kusma** [comment]
Right. I think that's a bad design-choice, we should instead dispatching the result to the JS threads in the end. Otherwise, we end up blocking the JS thread needlessly.

**Comment by kusma** [comment]
I've inspected the code, and I believe I've ruled out all possible call-sites from running JS on the wrong thread. The reason is HttpMessageHandlerRequest uses IDispatcher to dispatch its callbacks to the correct thread.

That being said, I think this part would have been easier to spot from the code if we dispatched manually. However, I just gave it an attempt, but the argument-marshalling due to different signatures would clearly have made this pretty awful, so I aborted the attempt.

In other words, I think this is as good as it gets for now.
